File: python/linked/linked_6.py
from linked_lists_1 import LinkedList, Node

# see Remove Duplicates for List and Node type definitions


# O(n)
def list_to_stack(l):
    stack = []
    node = l.head

    while node:
        stack.append(node)
        node = node.next

    return stack


# O(n)
def is_palindrome(l):
    list_node = l.head
    # O(n) create a stack to represent the list in reverse order
    stack = list_to_stack(l)
    list_length = len(stack)

    # O(n) compare the stack to the list
    while len(stack) >= list_length // 2:
        stack_node = stack.pop()
        if stack_node.value != list_node.value:
            return False
        list_node = list_node.next

    return True

File: ts/solutions/linked/linked_6.ts
import { LinkedList, Node_ } from './linked_1'

// see Remove Duplicates for List and Node type definitions

// O(n)
function listToStack(l: LinkedList): Node_[] {
  const stack: Node_[] = []
  let node = l.head
  while (node) {
    stack.push(node)
    node = node.next
  }
  return stack
}

// O(n)
function isPalindrome(l: LinkedList): boolean {
  let listNode: Node_ = l.head
  const stack = listToStack(l)
  const listLength = stack.length

  while (stack.length >= Math.floor(listLength / 2)) {
    const stackNode = stack.pop()
    if (stackNode.value !== listNode.value) {
      return false
    }
    listNode = listNode.next
  }

  return true
}
